data_structures/src/superblock.rs

```diff
@@ -544,6 +570,39 @@ fn magic_partition<T: Clone>(v: &[T], first: usize, size: usize) -> Vec<T> {
     v_subset
 }
 
+// Take size element out of v.len() starting with element at index first plus an offset:
+// magic_partition(v, 3, 3, r), v=[0, 1, 2, 3, 4, 5], r=[1].
+// Will return elements at index 4, 0, 2.
+fn magic_partition_2<T>(v: &[T], first: usize, size: usize, rand_distribution: &[u8]) -> HashSet<T>
+where
+    T: Clone + Eq + std::hash::Hash,
+{
+    let mut hs_subset = HashSet::default();
+    if first >= v.len() {
+        return hs_subset;
+    }
+    let each = v.len() / size;
+
+    let mut step_index = 0_usize;
+    let mut step = rand_distribution[step_index] as usize;
```

Comment:
Since we always access the `rand_distribution` variable sequentially, we can simplify the code by using iterators:

```suggestion
    let mut rand_gen = rand_distribution.iter().map(|x| usize::from(*x)).cycle();
    let mut step = rand_gen.next().unwrap();
```

This way we do not need the `step_index` variable. The unwrap is safe because `.cycle()` creates an infinite iterator that always returns something.

Reply:
Although this change will make it slightly more difficult to rewrite this function in solidity, so not sure if it's worth it.

Reply:
If the logic is the same, we could keep it simple